src/ui/PromotedResultsBadge/PromotedResultsBadge.ts

```diff
@@ -0,0 +1,187 @@
+import { Component } from '../../ui/Base/Component';
+import { IComponentBindings } from '../Base/ComponentBindings';
+import { ComponentOptions, Initialization, l, $$, ResultListEvents, Dom } from '../../Core';
+import { exportGlobally } from '../../GlobalExports';
+import { IQueryResult } from '../../rest/QueryResult';
+import { IDisplayedNewResultEventArgs } from '../../events/ResultListEvents';
+import 'styling/_PromotedResultsBadge';
+
+export interface IPromotedResultsBadgeOptions {
```

Comment:
That's the end user facing interface/type to determine options on the component. 

It's not an "internal" data structure that we want to "manipulate".

It needs to map one to one to an actual behaviour on the component, in a "flat" as straightforward manner. 

End user will actually be using this to configure the component in the page ( `data-color-for-recommended-results` , `data-color-for-featured-results` etc. ). And this will also be the options on the component presented in the interface editor.

Unless I am misunderstanding you, how would you propose the end user to specify markup attribute on the component to configure it ?

Also, just in general, adding an abstraction to something that is currently simple to prevent possible future complication sounds very over-engineered/premature optimization 😄 .